src/engraving/libmscore/slur.cpp

```diff
@@ -349,139 +366,327 @@ void SlurSegment::adjustEndpoints()
     }
 }
 
+void SlurSegment::avoidCollisions(PointF& pp1, PointF& p2, PointF& p3, PointF& p4, Transform& toSystemCoordinates, double& slurAngle)
+{
+    ChordRest* startCR = slur()->startCR();
+    ChordRest* endCR = slur()->endCR();
+    Segment* startSeg = startCR ? startCR->segment() : nullptr;
+    Segment* endSeg = endCR ? endCR->segment() : nullptr;
+    bool isCrossStaff = (startCR && endCR) ? startCR->vStaffIdx() != endCR->vStaffIdx() : false;
+
+    // Collision clearance at the center of the slur
+    double clearance = 0.75 * spatium(); // TODO: style
+    // balance: determines how much endpoint adjustment VS shape adjustment
+    // we will do (0 = only shape, 1 = only end point)
+    double balance = 0.4; // TODO: style (maybe?)
+
+    if (startSeg && endSeg && startSeg != endSeg && autoplace()) {
+        static constexpr unsigned maxIter = 30; // Max iterations allowed
+        const double vertClearance = slur()->up() ? clearance : -clearance;
+        const double step = slur()->up() ? -0.25 * spatium() : 0.25 * spatium();
+        // Divide slur in several rectangles to localize collisions
+        const unsigned npoints = 20;
+        std::vector<RectF> slurRects;
+        slurRects.reserve(npoints);
+        // Define separate collision areas (left-mid-center)
+        SlurCollision collision;
+
+        // CHECK FOR COLLISIONS
+        unsigned iter = 0;
+        do {
+            collision.left = collision.mid = collision.right = false;
```

Comment:
do you really need to set these values on every iteration?

Reply:
Yes, because at every iteration I modify the shape of the slur, so I must check for collisions again. Also, if I don't reset them I never exit the loop.

Reply:
then add a "reset" method into the structure

Reply:
ha, good point! done 👍